shared-bindings/bitmaptools/__init__.c

```diff
@@ -296,6 +296,62 @@ STATIC mp_obj_t bitmaptools_obj_fill_region(size_t n_args, const mp_obj_t *pos_a
 }
 
 MP_DEFINE_CONST_FUN_OBJ_KW(bitmaptools_fill_region_obj, 0, bitmaptools_obj_fill_region);
+//|
+//| def boundary_fill(
+//|        dest_bitmap: displayio.Bitmap,
+//|        x: int, y: int,
+//|        value: int, background_value: int) -> None:
+//|      """Draws the color value into the destination bitmap enclosed
+//|      area of pixels of the background_value color. Like "Paint Bucket"
+//|      fill tool.
+//|
+//|      :param bitmap dest_bitmap: Destination bitmap that will be written into
+//|      :param int x: x-pixel position of the first pixel to check and fill if needed
+//|      :param int y: y-pixel position of the first pixel to check and fill if needed
+//|      :param int value: Bitmap palette index that will be written into the
+//|             enclosed area in the destination bitmap
+//|      :param int background_value: Bitmap palette index that will filled with the
+//|             value color in the enclosed area in the destination bitmap"""
+//|      ...
+//|
+STATIC mp_obj_t bitmaptools_obj_boundary_fill(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    enum {ARG_dest_bitmap, ARG_x, ARG_y, ARG_value, ARG_background_value};
+
+    static const mp_arg_t allowed_args[] = {
+        {MP_QSTR_dest_bitmap, MP_ARG_REQUIRED | MP_ARG_OBJ},
+        {MP_QSTR_x, MP_ARG_REQUIRED | MP_ARG_INT},
+        {MP_QSTR_y, MP_ARG_REQUIRED | MP_ARG_INT},
+        {MP_QSTR_value, MP_ARG_REQUIRED | MP_ARG_INT},
+        {MP_QSTR_background_value, MP_ARG_REQUIRED | MP_ARG_INT},
+    };
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args, pos_args, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    displayio_bitmap_t *destination = MP_OBJ_TO_PTR(args[ARG_dest_bitmap].u_obj); // the destination bitmap
```

Comment:
Is there a way to check that this is a displayio_bitmap_t object? Probably unlikely but someone could pass a different object.

Reply:
I am not sure, but that is a good idea to try to handle it as gracefully as we can. Ill poke around some of the other core functions and see if there is anything in them that we could use in this one to try to guard against wrong arg tyles

Reply:
I think this function that Jeff showed me how to use is what we can use to ensure the argument is the right type:
```python
mp_obj_t mp_arg_validate_type(mp_obj_t obj, const mp_obj_type_t *type, qstr arg_name) {
    if (!mp_obj_is_type(obj, type)) {
        mp_raise_TypeError_varg(translate("%q must of type %q"), arg_name, type->name);
    }
    return obj;
}
```
I will give this a try tomorrow.

Reply:
Alright `6bd8a1d` adds a check to ensure the type of the Bitmap argument.

This code:
```
not_bitmap = {"hello": 123}
bitmaptools.boundary_fill(not_bitmap,5, 65, 2, 0)
```
Will now raise TypeError:
```
TypeError: dest_bitmap must of type Bitmap
```